lib/Partitioner/Partitioner.cpp

```diff
@@ -569,68 +581,390 @@ void Partitioner::doPartitioning(Function *F, NodeToFunctionMap &mapping) {
+    }
+    if (backendMap.find(backendKind) == backendMap.end()) {
+      BackendInfo backendInfo;
+      backendInfo.num = 1;
+      backendInfo.memSize = deviceInfo_[i].availableMemory;
+      backendInfo.backend =
+          hasBackends ? backends_[i] : createBackend(backendKind);
+      backendMap[backendKind] = backendInfo;
+      backends.push_back(backendMap[backendKind].backend);
+    } else {
+      backendMap[backendKind].num += 1;
+    }
+  }
+}
+
 llvm::Error Partitioner::Partition() {
-  // Find the representative function for running partitioning algorithm.
+  // Prepare the mapping between BackendKind and BackendInfo.
+  std::map<BackendKind, BackendInfo> backendMap;
+  std::vector<Backend *> backends;
+  getBackendMap(backendMap, backends);
+
+  // Step 0: Find the representative function for running partitioning
+  // algorithm.
   F_ = selectRepFunc(module_, memSize_);
-  uint64_t availMem = deviceInfo_[0].availableMemory;
-
-  if (memSize_ < availMem) {
-    // No partition is needed. Create DAGNode and return. This root is alway a
-    // dummy function.
-    for (auto F : module_->getFunctions()) {
-      std::unique_ptr<DAGNode> DAG0 = llvm::make_unique<DAGNode>();
-      DAG0->logicalDevices = {0};
-      DAG0->name = F->getName();
-      DAG0->module = module_;
-      std::unique_ptr<DAGNode> DAG1 = llvm::make_unique<DAGNode>();
-      DAG1->logicalDevices = {0};
-      DAG1->name = F->getName();
-      DAG1->parents.push_back(DAG0.get());
-      DAG0->children.push_back(DAG1.get());
-      DAGNodePtrVec nodes;
-      nodes.push_back(std::move(DAG1));
-      partitions_.push_back({std::move(DAG0), std::move(nodes)});
-    }
-    if (saturateHost_) {
-      // Saturate the Host.
-      saturateHost(1);
-    }
-    return llvm::Error::success();
-  }
-
-  // Prepare 1: Get the min memory usage for each op.
-  initOpMemUsage();
-
-  // Prepare 2: Get the roofline memory bandwidth estimate for each op.
-  initOpComputeTime();
-
-  // Partition
-  // Use BFS to do the initial partitioning. Starting from the final node, BFS
-  // until the memory limitation reached one by one.
-  NodeToFunctionMap partitionMap = selectPartitions(F_, availMem);
-
-  RETURN_ERR_IF_NOT(
-      partitionMap.getPartitions().size() <= deviceInfo_.size(),
-      llvm::formatv(
-          "Partition failed: the number of given devices ({0}) is fewer "
-          "than the required minimal partitions ({1}).",
-          deviceInfo_.size(), partitionMap.getPartitions().size())
-          .str());
-
-  doPartitioning(F_, partitionMap);
-
-  // Remove the original function after partitioning.
-  module_->eraseFunction(F_);
 
-  auto funcList = module_->getFunctions();
-  for (Function *F : funcList) {
-    (void)F;
-    assert(F->verify() && "Conversion led to invalid function");
+  // Step 1 : do the partition based on backends type.
+  FunctionToBackendKindMapTy funcToBackend;
+  std::string origName(F_->getName().data());
+  if (backends.size() == 1) {
```

Comment:
Is it necessary to special case when there is only 1 backend? Shouldn't the general case of multiple backends cover this case already? Of course, it would create just a single backend type based partition in this case, which is a waste. But it would work. 

Of course, if it is just a performance optimization, then it is fine. But then explain it in the comment.

Reply:
Yes, it is due to the performance. We don't want to do a function copy (which is a waste) since no function split is necessary